# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

---

**SOLUCIÓN**

Se puede empezar extrayendo información para obtener la función a optimizar.

Como se desea minimizar el precio del alimento, entonces nos vamos a la lista de precios por gramo de cada ingrediente y se concluye que la función a optimizar es la siguiente

$$
0.013x_1 + 0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6
$$

Además, se tienen diversas restricciones, a manera de ecuaciones:

- Peso total del alimento $\rightarrow$ $x_1 + x_2 + x_3 + x_4 + x_5 + x_6 = 100$

- Proteína $\rightarrow$ $0.1x_1 + 0.2x_2 + 0.15x_3 + 0.04x_5 \geq 8$

- Grasa $\rightarrow$ $0.08x_1 + 0.1x_2 + 0.11x_3 + 0.1x_4 + 0.15x_5 \geq 6$

- Fibra $\rightarrow$ $0.001x_1 + 0.005x_2 + 0.005x_3 + 0.1x_4 + 0.15x_5 \leq 2$

- Sal $\rightarrow$ $0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 \leq 0.4$

In [2]:
import numpy as np
import scipy.optimize as opt

In [3]:
c1 = np.array([0.013,0.008,0.01,0.002,0.005,0.001]) #Función a optimizar
A1 = np.array([[-0.1,-0.2,-0.15,0,-0.04,0],
            [-0.08,-0.1,-0.11,-0.01,-0.01,0],
            [0.001,0.005,0.005,0.1,0.15,0],
            [0.002,0.005,0.007,0.002,0.008,0]])
b1 = ([-8,-6,2,0.4])
A1eq = np.array([[1,1,1,1,1,1]])
b1eq = np.array([100])

In [4]:
opt.linprog(c1,A1,b1,A1eq,b1eq)

     con: array([3.05107051e-11])
     fun: 0.5200000000216561
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([4.00000000e+00, 7.70468134e-11, 1.70000000e+00, 1.00000000e-01])
  status: 0
 success: True
       x: array([4.50819667e-10, 6.00000000e+01, 4.08653405e-09, 2.29072691e-08,
       3.70730010e-10, 4.00000000e+01])

In [5]:
opt.linprog(c1,A1,b1,A1eq,b1eq, method = 'simplex')

     con: array([0.])
     fun: 0.52
 message: 'Optimization terminated successfully.'
     nit: 6
   slack: array([4. , 0. , 1.7, 0.1])
  status: 0
 success: True
       x: array([ 0., 60.,  0.,  0.,  0., 40.])

---

**RESULTADOS**

Para obtener el costo mínimo de $ \$0.52$ es necesario que el alimento contenga:
- $60g$ de Carne
- $40g$ de Gel

---

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en tres ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

---

**SOLUCIÓN**

Dentro de este problema existen diversas restricciones que se categorizan en 2 grupos:

1. Restricciones de oferta
    - $x_1 + x_2 + x_3 \leq 80$
    - $x_4 + x_5 + x_6 \leq 40$
    - $x_7 + x_8 + x_9 \leq 60$
    
2. Restricciones de demanda
    - $x_1 + x_4 + x_7 \geq 70$
    - $x_2 + x_5 + x_8 \geq 40$
    - $x_3 + x_6 + x_9 \geq 70$

Y la función a optimizar será la siguiente:

$$
5x_1 + 2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + x_8 + 2x_9
$$

In [6]:
c2 = np.array([5,2,7,3,6,6,6,1,2])
A2 = np.array([[1,1,1,0,0,0,0,0,0],
             [0,0,0,1,1,1,0,0,0],
             [0,0,0,0,0,0,1,1,1],
             [-1,0,0,-1,0,0,-1,0,0],
             [0,-1,0,0,-1,0,0,-1,0],
             [0,0,-1,0,0,-1,0,0,-1]])
b2 = np.array([80,40,60,-70,-40,-70])

In [7]:
opt.linprog(c2,A2,b2)

     con: array([], dtype=float64)
     fun: 539.9999999945555
 message: 'Optimization terminated successfully.'
     nit: 6
   slack: array([ 8.82408813e-10,  4.27498037e-10,  6.52505605e-10, -7.68807240e-10,
       -4.24719815e-10, -7.68878294e-10])
  status: 0
 success: True
       x: array([3.00000000e+01, 4.00000000e+01, 1.00000000e+01, 4.00000000e+01,
       4.66958755e-11, 6.07352053e-12, 2.20740129e-11, 2.37487168e-11,
       6.00000000e+01])

In [9]:
opt.linprog(c2,A2,b2, method = 'simplex')

     con: array([], dtype=float64)
     fun: 540.0
 message: 'Optimization terminated successfully.'
     nit: 9
   slack: array([0., 0., 0., 0., 0., 0.])
  status: 0
 success: True
       x: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

---

**RESULTADOS** 

Para obtener el costo mínimo de $540 u.m$ para satisfacer las necesidades de las tres ciudades, es necesario que:

- La Planta 1 distribuya 30 millones de kW a Guadalajara.
- La Planta 1 distribuya 40 millones de kW a León.
- La Planta 1 distribuya 10 millones de kW a Morelia.
- La Planta 2 distribuya 40 millones de kW a Guadalajara.
- La Planta 3 distribuya 60 millones de kW a Morelia.